manticore/ethereum/detectors.py

```diff
@@ -577,3 +577,111 @@ def did_evm_read_storage_callback(self, state, address, offset, value):
     def did_evm_write_storage_callback(self, state, address, offset, value):
         # concrete or symbolic write
         state.context.setdefault('{:s}.initialized_storage'.format(self.name), set()).add((address, offset))
+
+
+class DetectRaceCondition(Detector):
+    """Detects possible transaction race conditions (transaction order dependencies)"""
+    # At `setstoredaddress` in the `SSTORE` you taint the slot with 'taint_setstoredaddress'
+    # At `callstoredaddress` at the 'SLOAD' you read a value tainted with  'taint_setstoredaddress'
+    # (concrete or otherwise)
+    # at the `CALL` you check it taint(address) != 'taint_setstoredaddress'
+    # at the time of SSTORE or SLOAD the first 4 bytes in the CALLDATA must be constrained
+    # to a single solution / be concrete.
+    TAINT = 'written_storage_slots.'
+
+    def __init__(self, *a, **kw):
+        # Normally `add_finding_here` makes it unique reporting but
+        # we might try to report the same thing multiple times e.g. in consecutive instructions
+        # so we need to make our own 'unique findings' set too.
+        self.__findings = set()
+        import warnings
+        warnings.warn(
+            "The RaceCondition detector might not work properly for contracts that have only a fallback function. "
+            "See the detector's implementation and it's `_is_in_dispatcher` method for more information."
+        )
+        super().__init__(*a, **kw)
+
+    @staticmethod
+    def _is_in_dispatcher(state):
```

Comment:
You could cache this. Save this as a flag in state.context['in_dispatcher'] = {{None, True, False}}
So you do not need to re-solve once you already know.

Also the fallback and the dispatcher part are kind of indistinguishable this way. 
The fallback function is reached with the 4bytes beign diff from the known/defined functions. So at the fallback the 4bytes will have more than 2 solutions but never any of the unselected funcs.

Reply:
> Also the fallback and the dispatcher part are kind of indistinguishable this way.

Hmm... right. I think I will rename it to `_in_user_func` and make it return `return len(state.solve_n(state.platform.current_transaction.data[:4], 2)) == 1` instead.